### ISSUE TITLE
Add support for message sessions

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/NServiceBus.Transport.AzureServiceBus#535]
**Is your feature request related to a problem? Please describe.**
For a long time Particular has shied away from ordered messages delivery for historical reasons. The originally supported MSMQ transport did not provide guaranteed ordered delivery and the rest of the transports were molded into the assumption that ordered message processing is either not something that should be done, or accomplished using the saga feature.

More modern technologies such as Azure Service Bus are capable of handling messages in the order those were sent, providing a feature called [Message Sessions](https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-sessions).

**Describe the solution you'd like**
Add support for message sessions to allow endpoints that require message ordered processing at scale to do the job without resorting to more complex solutions such as sagas and state persistence.

**Additional context**
There are domains where message processing in the order those were sent is critical. Even the pizza [pizza](https://particular.net/blog/you-dont-need-ordered-delivery) preparation process.

**Comment by ramonsmits** [comment]
@SeanFeldman thanks for the suggestion but AFAIK message sessions are only providing ordering for a specific sender. If multiple senders. For example, sessions are useful in the context of multiple operations. Similar as in the past [NServiceBus supported message send batches](https://github.com/Particular/NServiceBus/blob/3.3.18/src/core/NServiceBus/IBus.cs#L92).

In the context of this feature request, are you talking about a message session that contains all outgoing message for a single incoming message. In other words, how would you want to see the life-time from the perspective of the sender?

**Comment by SeanFeldman** [comment]
> but AFAIK message sessions are only providing ordering for a specific sender.

That's incorrect. Different senders can contribute messages for a given session.

> For example, sessions are useful in the context of multiple operations. Similar as in the past [NServiceBus supported message send batches](https://github.com/Particular/NServiceBus/blob/3.3.18/src/core/NServiceBus/IBus.cs#L92).

No, @ramonsmits. That's not the same at all. I think you're mixing up messages grouped with Message Sessions. Have another look at the ASB-specific [feature](https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-sessions).

> In the context of this feature request, are you talking about a message session that contains all outgoing message for a single incoming message. In other words, how would you want to see the life-time from the perspective of the sender?

Senders are not the problem. Those can send messages and specify session ID. This, in theory, should be already possible with [native message access](https://docs.particular.net/transports/azure-service-bus/native-message-access#access-to-the-native-azure-service-bus-outgoing-message). Processing a single message should be done in the context of a session. Wherever there's a single message for that session or multiple messages, it doesn't matter. From NSB's perspective, it still receives one message at a time. It's _how_ it receives it needs to be mindful of whether the entity is session-enabled.

**Comment by tedvanderveen** [comment]
I think it's hard to over estimate the importance of adding support for ASB Sessions feature, as it is the only way to enable FIFO in ASB. In many scenarios correct handling of ordered events/messages is important. ASB Sessions offer a way to guarantee ordered processing and ensure that only a single queue/topic processor can process messages with the same SessionID.

**Comment by GaryFail** [comment]
Is there any update on this? Sessions on Azure Service Bus have been a life saver allowing us to scale out our receiving system, however the lack of support for this prevents us from using NServiceBus.

A simple example of the need for this would be a hospital admissions department admit a patient followed by quickly discharging them, two messages are published respectively, when only one instance of the consumer exists all is fine, but if we scale out to two or more instances in reality one could process the discharge first whilst the other is processing the admission with no knowledge of the discharge. This could result in a patient remained on the subscribing system as admitted.

The use of Azure Service Bus sessions and locking the session to the instance of the subscriber allows us to handle this but not if we move to NServiceBus, resulting in our inability to scale out.

**Comment by WilliamBZA** [comment]
No progress yet, but this feature request is on our radar, and was picked up recently for consideration but didn't quite make it to implementation then. We'll update this issue with progress when we start actively unpacking this.

**Comment by GaryFail** [comment]
Brilliant, thanks for the response William

**Comment by jdt** [comment]
Any more news on this? Does the fact that it was picked up for consideration mean we could have this feature in the foreseeable future?

**Comment by kbaley** [comment]
No new information yet. But we appreciate the comments. They help us determine what people are asking for to help with prioritization.

**Comment by lena-osterhus** [comment]
What is the status of this feature request?

We’re already using NServiceBus with Azure Service Bus transport in our Azure Function Apps and are running into issues when they are scaling out.

In short, an Azure Function App is scaling out, causing messages related to the same “session” (e.g., messages about the same Customer) to be processed in parallel across multiple instances. However, we need all messages concerning a given Customer to be processed sequentially by the same instance.

**Comment by DavidBoike** [comment]
While I can't promise that we are going to start working on this feature in the near future, I would like to get additional context on it, which can only help its hopes of getting prioritized. Right now it's too nebulous for that to be a realistic option.

I'd like to know how those participating in this thread would imagine sessions would/could be implemented in an NServiceBus context. From what I understand, a session is normally going to be associated with something temporary, as when @lena-osterhus mentioned a session being related to a single customer. So you could not configure an endpoint to only pay attention to a preconfigured session, as that would ignore all other messages not in that session, and wouldn't even be vaulable forever, as once that customer went away, no messages would apply.

I didn't review the nitty-gritty of how the Azure Service Bus transport works, but I want to say there is only one receiver that is grabbing messages and assigning them to worker threads until it hits the concurrency limit. It's not like for MaxConcurrency=8 there are 8 different receivers…the concurrency is managed by a Semaphore. So how would that have to change in a world where the transport supports sessions?

**Comment by SeanFeldman** [comment]
> From what I understand, a session is normally going to be associated with something temporary, as when [@lena-osterhus](https://github.com/lena-osterhus) mentioned a session being related to a single customer. 

Correct.

> So you could not configure an endpoint to only pay attention to a preconfigured session, as that would ignore all other messages not in that session, and wouldn't even be vaulable forever, as once that customer went away, no messages would apply.

Corect again. If someone needs a _specific_ session affinity, then NSB is not the right tool for them.


> I didn't review the nitty-gritty of how the Azure Service Bus transport works, but I want to say there is only one receiver that is grabbing messages and assigning them to worker threads until it hits the concurrency limit. It's not like for MaxConcurrency=8 there are 8 different receivers…the concurrency is managed by a Semaphore. So how would that have to change in a world where the transport supports sessions?

Multiple sessions can be processed concurrently. I don't know if anyone else's scenario would allow it or not, but you can have session A and session B being processed concurrently because what matters is that the messages for those sessions are processed by a single receiver in the order they were sent. I think that's the key point.

What it means for NSB is that the concurrency number would be translated into concurrent sessions a single instance of an endpoint can handle.

**Comment by DavidBoike** [comment]
Thanks @SeanFeldman! That provides more context that might allow a future Azure enhancement process to pick this up, but unfortunately I still can't promise any sort of timeline.

**Comment by ramonsmits** [comment]
> DLQ-ed message has to be resolved, which will "re-trigger" processing and deferred messages will be processed leveraging their sequence numbers in the session state.

@SeanFeldman does this require that specfic DLQ'd message to be moved back? Would that be problematic if that actual messages is removed from the DLQ, stored in in a database, and then retried?

**Comment by SeanFeldman** [comment]
> [@SeanFeldman](https://github.com/SeanFeldman) does this require that specfic DLQ'd message to be moved back? Would that be problematic if that actual messages is removed from the DLQ, stored in in a database, and then retried?

@ramonsmits what matters is that the next number in the sequence makes it for reprocessing. Wherever a message with that "id" is coming from a retried message in the DLQ or as a reconstructed message from the information in the DB, wouldn't matter much.

**Comment by ramonsmits** [comment]
> > [@SeanFeldman](https://github.com/SeanFeldman) does this require that specfic DLQ'd message to be moved back? Would that be problematic if that actual messages is removed from the DLQ, stored in in a database, and then retried?
> 
> @ramonsmits what matters is that the next number in the sequence makes it for reprocessing. Wherever a message with that "id" is coming from a retried message in the DLQ or as a reconstructed message from the information in the DB, wouldn't matter much.


@seanfeldman do you mean the native messageid property?

The transport intentionally sets this to a unique value to not be affected by deduplication is that is enabled. 

So if that is important that behavior must then be modified.

**Comment by SeanFeldman** [comment]
Not the ASB message ID. Sequential number of the message (also not SequenceNumber) that allows the processor to track that all messages have been processed.

**Comment by ramonsmits** [comment]
@SeanFeldman I think I now understand you.

> Sequential number of the message|

I thought this is a native ASB message property, but you mean custom sequence number added by the sender as ASB does not have anything native that guards against processing in a specific order.

Ideally a message processor adds a sequence number when it supports sessions. Having a sequence number in the payload helps to guard sequence in case a message gets DLQ'd. This helps to forward all remaining messages to DLQ too.

Then all messages from the DLQ can then be requeued it during staging with the right session identifier, and then use the embedded sequence number to re-queue it in the right sequence for most efficient processing?

**Comment by SeanFeldman** [comment]
Indeed.
